From 1a9153b92ed68183e0c75703a4c95c044449d3e7 Mon Sep 17 00:00:00 2001
From: xvhaocheng <xvhaocheng@gmail.com>
Date: Sat, 17 Oct 2026 14:52:05 +0900
Subject: [PATCH 18/27] Allow research profiles to omit legacy E3 trace
 hardware

Source-organization: 2026-10-17; new commit authored at actual current time.
Feature-ID: CVA6-LEGACY-E3-SELECT
Original-source: 8c1e08b3
Validation scope: source/patch identity only; no build, simulation or experiment.
---
 .../fpga/src/ariane_peripherals_xilinx.sv     | 43 +++++++++++--------
 corev_apu/fpga/src/ariane_xilinx.sv           |  4 ++
 2 files changed, 29 insertions(+), 18 deletions(-)

diff --git a/corev_apu/fpga/src/ariane_peripherals_xilinx.sv b/corev_apu/fpga/src/ariane_peripherals_xilinx.sv
index 207d86a0edf86dee5b698eacda0621fca7fec44e..df2ffba401f0efcf2282cb8ecb0de158b6b192a4 100644
--- a/corev_apu/fpga/src/ariane_peripherals_xilinx.sv
+++ b/corev_apu/fpga/src/ariane_peripherals_xilinx.sv
@@ -29,6 +29,7 @@ module ariane_peripherals #(
     parameter bit InclTimer    =  1,
     parameter bit InclDMA      =  0,
     parameter bit InclIOMMU    =  0,
+    parameter bit IOMMULegacyE3Observer = 1,
     parameter int unsigned IOMMUDSFixedDelayInput  = 0,
     parameter int unsigned IOMMUDSFixedDelayOutput = 0
 ) (
@@ -941,7 +942,8 @@ module ariane_peripherals #(
                 (axi_dma_tr_resp[o].b_valid && |axi_dma_tr_resp[o].b.id);
         end
         dma_observer #(.NumPorts(NumDMADevices), .PathEnable(InclIOMMU),
-                       .E3Enable(InclIOMMU), .ControlledEnable(InclIOMMU),
+                       .E3Enable(InclIOMMU && IOMMULegacyE3Observer),
+                       .ControlledEnable(InclIOMMU && IOMMULegacyE3Observer),
                        .P4MEnable(InclIOMMU), .ResearchEnable(InclIOMMU),
                        .AxiIdWidth(ariane_soc::IdWidth)) i_dma_observer (
             .clk_i, .rst_ni,
@@ -1033,23 +1035,28 @@ module ariane_peripherals #(
             assign launch_cfg_rdata = '0;
         end
 
-        dma_e3_observer i_e3_observer (
-            .clk_i,.rst_ni,.arm_i(e3_control[1]),.clear_i(e3_control[2]),
-            .active_i(e3_control[3]),.open_i(e3_control[4]),.close_i(e3_control[5]),
-            .abort_i(e3_control[6]),.delay128_i(e3_control[7]),
-            .ctl_i(e3_ctl),.tl_i(e3_tl),.index_i(e3_index),.data_o(e3_data),.error_o(e3_error),
-            .tr_fire_i({axi_iommu_tr_req.aw_valid && axi_iommu_tr_resp.aw_ready,
-                        axi_iommu_tr_req.ar_valid && axi_iommu_tr_resp.ar_ready}),
-            .tr_id_i('{axi_iommu_tr_req.ar.id,axi_iommu_tr_req.aw.id}),
-            .tr_address_i('{64'(axi_iommu_tr_req.ar.addr),64'(axi_iommu_tr_req.aw.addr)}),
-            .out_valid_i({axi_iommu_comp_req.aw_valid,axi_iommu_comp_req.ar_valid}),
-            .out_ready_i({axi_iommu_comp_resp.aw_ready,axi_iommu_comp_resp.ar_ready}),
-            .out_id_i('{axi_iommu_comp_req.ar.id,axi_iommu_comp_req.aw.id}),
-            .return_i({axi_iommu_comp_resp.b_valid && axi_iommu_comp_req.b_ready,
-                       axi_iommu_comp_resp.r_valid && axi_iommu_comp_resp.r.last && axi_iommu_comp_req.r_ready}),
-            .return_id_i('{axi_iommu_comp_resp.r.id,axi_iommu_comp_resp.b.id}),
-            .wlast_i(axi_iommu_comp_req.w_valid && axi_iommu_comp_req.w.last && axi_iommu_comp_resp.w_ready)
-        );
+        if (IOMMULegacyE3Observer) begin : gen_e3_observer
+            dma_e3_observer i_e3_observer (
+                .clk_i,.rst_ni,.arm_i(e3_control[1]),.clear_i(e3_control[2]),
+                .active_i(e3_control[3]),.open_i(e3_control[4]),.close_i(e3_control[5]),
+                .abort_i(e3_control[6]),.delay128_i(e3_control[7]),
+                .ctl_i(e3_ctl),.tl_i(e3_tl),.index_i(e3_index),.data_o(e3_data),.error_o(e3_error),
+                .tr_fire_i({axi_iommu_tr_req.aw_valid && axi_iommu_tr_resp.aw_ready,
+                            axi_iommu_tr_req.ar_valid && axi_iommu_tr_resp.ar_ready}),
+                .tr_id_i('{axi_iommu_tr_req.ar.id,axi_iommu_tr_req.aw.id}),
+                .tr_address_i('{64'(axi_iommu_tr_req.ar.addr),64'(axi_iommu_tr_req.aw.addr)}),
+                .out_valid_i({axi_iommu_comp_req.aw_valid,axi_iommu_comp_req.ar_valid}),
+                .out_ready_i({axi_iommu_comp_resp.aw_ready,axi_iommu_comp_resp.ar_ready}),
+                .out_id_i('{axi_iommu_comp_req.ar.id,axi_iommu_comp_req.aw.id}),
+                .return_i({axi_iommu_comp_resp.b_valid && axi_iommu_comp_req.b_ready,
+                           axi_iommu_comp_resp.r_valid && axi_iommu_comp_resp.r.last && axi_iommu_comp_req.r_ready}),
+                .return_id_i('{axi_iommu_comp_resp.r.id,axi_iommu_comp_resp.b.id}),
+                .wlast_i(axi_iommu_comp_req.w_valid && axi_iommu_comp_req.w.last && axi_iommu_comp_resp.w_ready)
+            );
+        end else begin : gen_no_e3_observer
+            assign e3_data = '0;
+            assign e3_error = 1'b0;
+        end
 
         AXI_BUS #(
             .AXI_ADDR_WIDTH ( AxiAddrWidth             ),
diff --git a/corev_apu/fpga/src/ariane_xilinx.sv b/corev_apu/fpga/src/ariane_xilinx.sv
index 8877441d151a8d334f9dd5ea49af5234afbdcc7c..748d677bb5a45545a23aa748afce1426510eec40 100644
--- a/corev_apu/fpga/src/ariane_xilinx.sv
+++ b/corev_apu/fpga/src/ariane_xilinx.sv
@@ -164,6 +164,9 @@ localparam AxiUserWidth = ariane_pkg::AXI_USER_WIDTH;
 // At 50 MHz, 100 cycles model about 2 us of additional remote-memory latency.
 localparam int unsigned IOMMUDSFixedDelayInput = 50;
 localparam int unsigned IOMMUDSFixedDelayOutput = 50;
+// Legacy E3 tracing remains enabled unless a profile explicitly disables it.
+// The research profile disables this module to recover FPGA capacity.
+localparam int unsigned IOMMULegacyE3Observer = 1;
 
 `AXI_TYPEDEF_ALL(axi_slave,
                  logic [    AxiAddrWidth-1:0],
@@ -812,6 +815,7 @@ ariane_peripherals #(
     .InclIOMMU    ( 1'b1             ),
     .IOMMUDSFixedDelayInput  ( IOMMUDSFixedDelayInput  ),
     .IOMMUDSFixedDelayOutput ( IOMMUDSFixedDelayOutput ),
+    .IOMMULegacyE3Observer   ( IOMMULegacyE3Observer   ),
     `ifdef KINTEX7
     .InclSPI      ( 1'b1         ),
     .InclEthernet ( 1'b1         )
